File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
   public int[] maxDepth(TreeNode root) {
        if (root == null) {
            return new int[] {0, 0};
        }
        
        int[] left = maxDepth(root.left);
        int[] right = maxDepth(root.right);
        
        int height = 1 + Math.max(left[0], right[0]);
        int diameter = Math.max(left[1], Math.max(right[1], left[0] + right[0]));
        
        return new int[] {height, diameter};
    }

    public int diameterOfBinaryTree(TreeNode root) {
        return maxDepth(root)[1];
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        diameter = [0]
        def maxDepth(root):
            
            if root == None:
                return 0
            
            left = maxDepth(root.left)
            right = maxDepth(root.right)

            diameter[0] = max(left+right,diameter[0])

            return 1+max(left, right)

        l = maxDepth(root)

        return diameter[0]
                
        